### PR TITLE
Optimize sync result code

### PR DESCRIPTION
**Pull request by leplatrem**
Before:
https://perfht.ml/2Xthz1c

After:
https://perfht.ml/2WVRfsq

--> Time spent in `importChanges()` becomes ridiculous.

### CODE REVIEW

File: src/collection.js [hunk 1]
```diff
@@ -35,8 +35,6 @@ export class SyncResultObject {
    */
   static get defaults() {
```
Comment (glasserc) [inline src/collection.js]: Can we rename this to something that is clearly list-specific?

File: test/collection_test.js [hunk 2]
```diff
@@ -2266,6 +2281,18 @@ describe("Collection", () => {
...
+            lastModified: result.lastModified,
+            errors: result.errors,
+            created: result.created,
+            updated: result.updated,
+            deleted: result.deleted,
+            skipped: result.skipped,
+            published: result.published,
+            conflicts: result.conflicts,
+            resolved: result.resolved,
+          }))
```
Comment (glasserc) [inline test/collection_test.js]: Maybe we should extract this as a utility function, something like `reifySyncResult` ?

**Comment by glasserc** [comment]
Are these profiles the right ones? I remember you showing me one that has lots of `addOneRecord` calls, but this one doesn't seem to have any..

**Comment by leplatrem** [comment]
> Are these profiles the right ones? I remember you showing me one that has lots of addOneRecord calls, but this one doesn't seem to have any..

You're right... they don't show the right info.

Here is a new try:

Before https://perfht.ml/2ZN8M7E 
After https://perfht.ml/2FxY6SP

File: src/collection.js [hunk 3]
```diff
@@ -69,44 +61,100 @@ export class SyncResultObject {
...
    */
   reset(type) {
-    this[type] = SyncResultObject.defaults[type];
-    this.ok = this.errors.length + this.conflicts.length === 0;
+    this._lists[type] = [];
+    delete this._cached[type];
     return this;
   }
+
+  toObject() {
```
Comment (glasserc) [inline src/collection.js]: Maybe add a comment that this is just for testing?

**Comment by leplatrem** [comment]
> I guess we no longer deduplicate errors or conflicts, although I imagine that isn't really important.

I don't think we did before:

https://github.com/Kinto/kinto.js/blob/ad789ca33256450c98e30dbff7fb6b3914117a0a/src/collection.js#L78-L79

**Pull request merged**